user-select: initial content within user-select: none should be copied

rniwa opened the pull request:
#### 3837614ab70da684405b3f6f704694b6a5ff6bf5
<pre>
user-select: initial content within user-select: none should be copied
<a href="https://bugs.webkit.org/show_bug.cgi?id=249507">https://bugs.webkit.org/show_bug.cgi?id=249507</a>

Reviewed by Darin Adler.

Allow content inside user-select: none to be copyable if it&apos;s not user-select: none.

The fix is simple for TextIterator where we can set m_handledChildren to be false when
ignoring user-select: none content so that if there is any visible content we won&apos;t skip.

The fix is more involved for StyledMarkupAccumulator (for HTML) and HTMLConverter (RTFD).
These classes need to decide whether the whole element should be skipped or not based on
if an element with user-select: none contains any content that is not user-select: none or not.

If it does, then we need to traverse through these elements with appropriate block/paragraph
styling and emit string for those content that is not user-select: none.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-expected.txt:
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none.html:

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::advance):

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_enterElement):
(HTMLConverter::_processText):

* Source/WebCore/editing/markup.cpp:
(WebCore::UserSelectNoneStateCache::UserSelectNoneStateCache): Added.
(WebCore::UserSelectNoneStateCache::parentNode): Added.
(WebCore::UserSelectNoneStateCache::firstChild): Added.
(WebCore::UserSelectNoneStateCache::nextSibling): Added.
(WebCore::UserSelectNoneStateCache::computeState): Added.
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):

* Source/WebCore/editing/markup.h:
(WebCore::UserSelectNoneStateCache::nodeOnlyContainsUserSelectNone): Added.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(CopyRTF.StripsUserSelectNone):

Canonical link: <a href="https://commits.webkit.org/258054@main">https://commits.webkit.org/258054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8c58b06845472f8949a34172efb5d5cccae0c13

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100728 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110027 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170301 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104716 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/400 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93129 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107873 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106511 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91416 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34774 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77745 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3566 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24332 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3590 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/398 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9701 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43829 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5368 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2895 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/editing/markup.cpp:
```diff
@@ -223,6 +223,76 @@ String sanitizeMarkup(const String& rawHTML, MSOListQuirks msoListQuirks, std::o
+
+ContainerNode* UserSelectNoneStateCache::parentNode(Node& node)
+{
+    if (UNLIKELY(m_useComposedTree))
```
Is this code so hot we need to explicitly use `UNLIKELY`?

rniwa replied:
Actually, we probably don't need LIKELY/UNLIKELY here.

darinadler commented on Source/WebCore/editing/markup.h:
```diff
@@ -59,6 +59,24 @@ enum class MSOListQuirks : bool { CheckIfNeeded, Disabled };
 
+class UserSelectNoneStateCache {
+public:
+    UserSelectNoneStateCache(TreeType);
```
`explicit`

rniwa replied:
Will fix.

darinadler commented on Source/WebCore/editing/markup.h:
```diff
@@ -59,6 +59,24 @@ enum class MSOListQuirks : bool { CheckIfNeeded, Disabled };
+    Node* firstChild(Node&);
+    Node* nextSibling(Node&);
+
+    enum class State { NotUserSelectNone, Mixed, OnlyUserSelectNone };
```
Maybe `: uint8_t`?

rniwa replied:
Sure.

rniwa committed:
user-select: initial content within user-select: none should be copied
https://bugs.webkit.org/show_bug.cgi?id=249507

Reviewed by Darin Adler.

Allow content inside user-select: none to be copyable if it's not user-select: none.

The fix is simple for TextIterator where we can set m_handledChildren to be false when
ignoring user-select: none content so that if there is any visible content we won't skip.

The fix is more involved for StyledMarkupAccumulator (for HTML) and HTMLConverter (RTFD).
These classes need to decide whether the whole element should be skipped or not based on
if an element with user-select: none contains any content that is not user-select: none or not.

If it does, then we need to traverse through these elements with appropriate block/paragraph
styling and emit string for those content that is not user-select: none.

* LayoutTests/editing/pasteboard/copy-content-with-user-select-none-expected.txt:
* LayoutTests/editing/pasteboard/copy-content-with-user-select-none.html:

* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::advance):

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(HTMLConverter::HTMLConverter):
(HTMLConverter::_enterElement):
(HTMLConverter::_processText):

* Source/WebCore/editing/markup.cpp:
(WebCore::UserSelectNoneStateCache::UserSelectNoneStateCache): Added.
(WebCore::UserSelectNoneStateCache::parentNode): Added.
(WebCore::UserSelectNoneStateCache::firstChild): Added.
(WebCore::UserSelectNoneStateCache::nextSibling): Added.
(WebCore::UserSelectNoneStateCache::computeState): Added.
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):

* Source/WebCore/editing/markup.h:
(WebCore::UserSelectNoneStateCache::nodeOnlyContainsUserSelectNone): Added.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/CopyRTF.mm:
(CopyRTF.StripsUserSelectNone):

Canonical link: https://commits.webkit.org/258054@main
